src/circuit/circuit_frame.rs

```diff
@@ -1344,352 +1338,457 @@ fn reduce_cons<F: LurkField, CS: ConstraintSystem<F>>(
             &expanded2,
         )?;
 
-        let continuation_letrec = AllocatedContPtr::construct(
-            &mut cs_letrec.namespace(|| "letrec continuation"),
-            store,
-            &g.letrec_cont_tag,
-            &[&var, &expanded_, env, cont],
-        )?;
-
         let output_expr = AllocatedPtr::pick(
             &mut cs_letrec.namespace(|| "pick body1 or val"),
             &bindings_is_nil,
             &body1,
             &val,
         )?;
 
-        let output_cont_let = AllocatedContPtr::pick(
-            &mut cs_letrec.namespace(|| "pick cont or newer let"),
-            &bindings_is_nil,
-            cont,
-            &continuation_let,
-        )?;
-
-        let output_cont_letrec = AllocatedContPtr::pick(
-            &mut cs_letrec.namespace(|| "pick cont or newer letrec"),
-            &bindings_is_nil,
-            cont,
-            &continuation_letrec,
-        )?;
-
         let the_expr = AllocatedPtr::pick(
             &mut cs_letrec.namespace(|| "the_expr_let"),
             &cond_error,
             expr,
             &output_expr,
         )?;
 
-        let the_cont_let = AllocatedContPtr::pick(
-            &mut cs_letrec.namespace(|| "the_cont_let"),
-            &cond_error,
-            &g.error_ptr_cont,
-            &output_cont_let,
+        (
+            the_expr,
+            var,
+            expanded,
+            expanded_,
+            bindings_is_nil,
+            cond_error,
+        )
+    };
+
+    // head == LET and LETREC
+    /////////////////////////////////////////////////////////////////////////////
+    let let_continuation_components: &[&dyn AsAllocatedHashComponents<F>; 4] =
+        &[&var_letrec, &expanded_let, env, cont];
```

Comment:
I think `var_letrec` is actually shared between the `let` and `letrec` cases. In that case, the name is probably wrong and will cause confusion later.